### PR TITLE
#20733 adding the role.key.substitution functionality

### LINKED ISSUE #20773: SAML - Allow expression substitution from SAML roles mapped to dotCMS roles by role key
When SAML authentication is configured in dotCMS and roles are inherited from IdP, dotCMS follows this approach for roles that come in SAML assertion attributes (`build.roles` property is set to `all` or `idp`):

1. If there is no dotCMS role with the same role key as the role name sent from IdP, a new dotCMS role will be created and its key will be set to the role name sent in the assertion attribute.
2. But it dotCMS already has a role with the same role key as the role name sent from IdP, then dotCMS won't create a new role but it will assign the matching role to the user (in case the user has not been assigned to that role yet)

This approach allows mapping IdP roles with existing dotCMS roles. For example, I can assign an existing dotCMS role as 'CMS Administrator' to a user automatically when the user logs in for the first time. However the IdP can have some constraints in role names, like not allowing spaces, so I can't send the role 'CMS Administrator' in the assertion attribute. With some substitution mechanisms enabled, I would be able to send a role name as 'CMS_sepsep_Administrator' and dotCMS will replace it with the appropriate role key 'CMS Administrator'. 

### Proposed Solution
A new configuration property for the SAML plugin will allow adding a substitution expression to use when mapping dotCMS role keys. For example:
```
role.key.substitution=/_sepsep_/ /
```
That substitution expression will replace the string `_sepsep_` with white space before matching role name from IdP against role key from dotCMS.

### PR DESCRIPTION
**Pull request by jdotcms**
A new configuration property for the SAML plugin will allow adding a substitution expression to use when mapping dotCMS role keys. For example:

role.key.substitution=/_sepsep_/ /

That substitution expression will replace the string _sepsep_ with white space before matching role name from IdP against role key from dotCMS.

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotcms/auth/providers/saml/v1/SAMLHelperTest.java [hunk 1]
```diff
@@ -220,4 +226,50 @@ public void testResolveUser_byEmailAddress() throws DotDataException, DotSecurit
         Assert.assertNotEquals(nativeUser.getFirstName(),  recoveredNativeUser.getFirstName());
         Assert.assertEquals (nativeUser.getLastName(),     recoveredNativeUser.getLastName());
     }
+
+    @Test()
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/auth/providers/saml/v1/SAMLHelperTest.java:230]: doc

File: dotCMS/src/integration-test/java/com/dotcms/auth/providers/saml/v1/SAMLHelperTest.java [hunk 2]
```diff
@@ -220,4 +226,50 @@ public void testResolveUser_byEmailAddress() throws DotDataException, DotSecurit
...
+        Assert.assertFalse(substitutionTokenOpt.isPresent());
+        final String expected     = "CMS Administrator";
+
+        final String roleResult   = samlHelper.processReplacement(expected, substitutionTokenOpt);
+
+        Assert.assertEquals(expected, roleResult);
+    }
+
+    @Test()
+    public void test_getRoleKeySubstitution_processReplacement_doing_substitution() throws DotDataException, DotSecurityException, IOException {
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/auth/providers/saml/v1/SAMLHelperTest.java:253]: doc

File: dotCMS/src/test/java/com/dotmarketing/util/RegEXTest.java [hunk 3]
```diff
@@ -82,4 +86,42 @@ public void testCaseUnSensitiveAndCaseSensitiveMixRegex () throws ParseException
...
+        final String token        = "CMS_sepsep_Administrator";
+        final String expected     = "CMS Administrator";
+
+        final String result = RegEX.replace(token, replacement, replace);
+
+        Assert.assertEquals(expected, result);
+    }
+
+    @Test
+    public void test_replace_token_by_white_space_on_pattern () throws ParseException {
```
Comment (freddyucv) [inline dotCMS/src/test/java/com/dotmarketing/util/RegEXTest.java:104]: doc

File: dotCMS/src/test/java/com/dotmarketing/util/RegEXTest.java [hunk 4]
```diff
@@ -82,4 +86,42 @@ public void testCaseUnSensitiveAndCaseSensitiveMixRegex () throws ParseException
         Assert.assertTrue(RegEX.containsCaseInsensitive(text2, regex));
         Assert.assertTrue(RegEX.containsCaseInsensitive(text3, regex));
     }
+
+    @Test
+    public void test_replace_token_by_white_space () throws ParseException {
```
Comment (freddyucv) [inline dotCMS/src/test/java/com/dotmarketing/util/RegEXTest.java:91]: doc

**Review by freddyucv — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Pull request merged**